### ISSUE TITLE
update dependencies in "leaf" packages of a yarn/lerna monorepo

### ISSUE THREAD

**Issue by mmkal** [renovatebot/config-help#752]
### Which Renovate are you using?

WhiteSource Renovate App

### Which platform are you using?

GitHub.com

### Have you checked the logs? Don't forget to include them if relevant

Yes, here's an example: https://app.renovatebot.com/dashboard#github/mmkal/ts/202802903

### What would you like to do?

Have renovate update dependencies in "leaf" packages of a yarn/lerna monorepo.

As an example. the dependency fp-ts is on 2.0.1 in the package io-ts-extra: [link](https://github.com/mmkal/ts/blob/bd2e2fe4d791050cbbc59177f3935612705152bb/packages/io-ts-extra/package.json#L17)

And it's on 2.1.0 in yarn.lock: [link](https://github.com/mmkal/ts/blob/bd2e2fe4d791050cbbc59177f3935612705152bb/yarn.lock#L4491-L4494)

But 2.7.0 is available on npm: https://www.npmjs.com/package/fp-ts/v/2.7.0

Is there a config switch I need to enable to tell renovate that I would like it to update the dependencies in leaf packages too? In the logs linked above, renovate isn't detecting any updates (there are no PRs, and the logs show `DEBUG: 0 flattened updates found`). It _has_ worked for devDependencies, which are in the root package.json ([example](https://github.com/mmkal/ts/pull/76)). But once it's up to date with them, it doesn't seem to work for the "leaves". Would be great to understand how I can tell it I'd like those updated too.

Some more info:
- renovate is correctly interpreting the project as a monorepo, and it's aware of the fp-ts dependency (it finds the current version just above the `"packageFile": "packages/io-ts-extra/package.json"` line)
- fp-ts is just an example of a "leaf" dependency - there are more, which you can see from the result of `yarn outdated` on the command line:

```
Package         Current Wanted  Latest  Workspace             Package Type    URL
@babel/core     7.9.0   7.10.4  7.10.4  eslint-plugin-codegen dependencies    https://babeljs.io/
@babel/parser   7.9.4   7.10.4  7.10.4  eslint-plugin-codegen dependencies    https://babeljs.io/
@babel/traverse 7.9.0   7.10.4  7.10.4  eslint-plugin-codegen dependencies    https://babeljs.io/
@types/node     12.11.1 12.11.1 14.0.23 ts                    devDependencies https://github.com/DefinitelyTyped/DefinitelyTyped.git
expect          26.0.1  26.1.0  26.1.0  eslint-plugin-codegen dependencies    https://github.com/facebook/jest#readme
fp-ts           2.1.0   2.7.0   2.7.0   eslint-plugin-codegen dependencies    https://github.com/gcanti/fp-ts
fp-ts           2.1.0   2.7.0   2.7.0   io-ts-extra           dependencies    https://github.com/gcanti/fp-ts
glob            7.1.4   7.1.6   7.1.6   eslint-plugin-codegen dependencies    https://github.com/isaacs/node-glob#readme
io-ts           2.2.4   2.2.8   2.2.8   eslint-plugin-codegen dependencies    https://github.com/gcanti/io-ts
io-ts           2.0.1   2.2.8   2.2.8   io-ts-extra           dependencies    https://github.com/gcanti/io-ts
lodash          4.17.15 4.17.19 4.17.19 eslint-plugin-codegen dependencies    https://lodash.com/
```

**Comment by viceice** [comment]
you have `^2.1.0` in your `package.json` which is satisfies `2.7.0`, so no need todo something by renovate.

If you like to change the behavior, you need to configure [`raangeStrategy`](https://docs.renovatebot.com/configuration-options/#rangestrategy)

**Comment by mmkal** [comment]
Sorry, should have clarified - I was expecting yarn.lock to update, rather than package.json. That's still on the old version too,

**Comment by viceice** [comment]
`rangeStrategy='update-lockfile'` ? See linked documentation above.

**Comment by mmkal** [comment]
@viceice thank you! Only snag is I'd still like to pin devDependencies - would something like this work? I ask because `rangeStrategy` doesn't appear under `packageRules` in https://docs.renovatebot.com/renovate-schema.json:

```json
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": [
    "config:base",
    ":semanticCommits"
  ],
  "gitAuthor": "bot@renovateapp.com",
  "prConcurrentLimit": 3,
  "packageRules": [
    {
      "depTypeList": ["dependencies"],
      "rangeStrategy": "update-lockfile"
    },
    {
      "depTypeList": ["devDependencies"],
      "automerge": true
    },
    {
      "depTypeList": ["devDependencies"],
      "packagePatterns": [".*eslint.*"],
      "groupName": "eslint packages",
      "labels": ["linting"]
    }
  ]
}
```

**Comment by viceice** [comment]
@mmkal Should work, the schema is only a help, but is not always fully correct because of some inheritance tradeof

**Comment by mmkal** [comment]
Great - thanks v much for such a fast response!

**Comment by mmkal** [comment]
@viceice reopening - something still isn't quite working. The config is now updated to more or less the above - [link to renovate.json in github](https://github.com/mmkal/ts/blob/main/renovate.json).

Based on this config, I'd expect the fp-ts entry in the lockfile to have a PR opened for it, but nothing happened. Here's a recent renovate log: https://app.renovatebot.com/dashboard#github/mmkal/ts/203278869

It evaluates several branches, including one called `renovate/fp-ts-2.x-lockfile`, as expected. It knows that the latest version is 2.7.0 - it logs `findPr(renovate/fp-ts-2.x-lockfile, fix(deps): update dependency fp-ts to v2.7.0, !open)`. But a few lines later, after the `docker run` command, it says `yarn.lock hasn't changed`, so doesn't open a PR.

It seems something is still off about the config - pasted below for convenience/in case main changes:

<details>
<summary>renovate.json</summary>

```json

{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": [
    "config:base",
    ":semanticCommits"
  ],
  "prConcurrentLimit": 3,
  "packageRules": [
    {
      "depTypeList": ["dependencies"],
      "rangeStrategy": "update-lockfile"
    },
    {
      "depTypeList": ["dependencies"],
      "updateTypes": ["lockFileMaintenance"],
      "automerge": true
    },
    {
      "depTypeList": ["devDependencies"],
      "automerge": true
    },
    {
      "depTypeList": ["devDependencies"],
      "packagePatterns": ["eslint"],
      "groupName": "eslint packages",
      "labels": ["linting"]
    }
  ]
}
```
</details>

**Comment by mmkal** [comment]
After some more investigation: I think it's a yarn bug.

I tried running the `yarn ugprade` command that `docker run` is using, which was `yarn upgrade fp-ts`. It didn't do anything. But, if I run `yarn upgrade-interactive fp-ts` then manually select the available versions, fp-ts does update - but it comes along with a few other changes, which effectively undo the effects of `resolutions` (which are needed to vulnerabilities in some other packages). If you _then_ run `yarn upgrade fp-ts` it re-fixes the resolutions. I pushed a branch so it's possible to see the result of each command as a git commit:

- result of `yarn upgrade fp-ts`: none
- result of `yarn upgrade-interactive fp-ts` and selecting both options: https://github.com/mmkal/ts/commit/02f29f91dd9e26a62c012c7d7a8436b28221697f
- result of `yarn upgrade fp-ts` after that: https://github.com/mmkal/ts/commit/2627490bca66220cdf1e76ff60416e213047a34e

Note that [the diff in the end](https://github.com/mmkal/ts/compare/0b96e7abf2bee2a6f73bc200f31326aed2189894...2627490bca66220cdf1e76ff60416e213047a34e) is what renovate is aiming for.

I might be missing something here, but it seems like a bug. I will get a bit more info and open a yarn issue, but I imagine it won't be fixed in yarn v1 since v2 is on the way. So in the meantime, I think it'd be worth working around this in renovate. And there is a single-command, non-interactive, fairly reasonably workaround. If, instead of running `yarn upgrade fp-ts` in the `docker run` command, renovate runs `yarn upgrade fp-ts@^2.1.0` (where `^2.1.0` is the range found in package.json), everything works as expected. That dependency is updated, and there are no strange side-effects.

**Comment by viceice** [comment]
@mmkal This looks really like a yarn bug.

I think yarn simply does nothing, because the current lockfile satisfies all contraints. Maybe a better solution for you is to use the `bump` `rangeStrategy`

@rarkins Do you think we should fix that in renovate? Is it easily fixable for us?

**Comment by rarkins** [comment]
I don't think we should work around this if it's a yarn bug.

**Comment by mmkal** [comment]
@viceice @rarkins I opened an issue against yarn (linked above), but I'm not particularly hopeful it'll be fixed, based on [this issue](https://github.com/yarnpkg/yarn/issues/7234) which sounds similar in severity/maybe cause, and has had no response of any kind for over a year.

So I'm very much hoping for a workaround - otherwise renovate just won't work for yarn projects that use resolutions. To that end, I also opened a shot-in-the-dark PR (also linked above) with a workaround that was effective for me, and that felt reasonable and easy enough to maintain.
